src/geointerface.jl

```diff
@@ -23,57 +35,282 @@ let pointtypes = (wkbPoint, wkbPoint25D, wkbPointM, wkbPointZM),
+        elseif gtype == wkbCurvePolygon
+            GeoInterface.CurvePolygonTrait()
+        elseif gtype == wkbMultiSurface
+            GeoInterface.MultiSurfaceTrait()
+        elseif gtype == wkbPolyhedralSurface
+            GeoInterface.PolyhedralSurfaceTrait()
+        elseif gtype == wkbTIN
+            GeoInterface.TINTrait()
+        elseif gtype == wkbTriangle
+            GeoInterface.TriangleTrait()
         else
-            @warn "unknown geometry type" gtype
-            :Unknown
+            @warn "unknown geometry type"
+            nothing
         end
     end
 
-    function GeoInterface.coordinates(g::AbstractGeometry)
-        gtype = getgeomtype(g)
-        ndim = getcoorddim(g)
-        return if gtype in pointtypes
-            if ndim == 2
-                Float64[getx(g, 0), gety(g, 0)]
-            elseif ndim == 3
-                Float64[getx(g, 0), gety(g, 0), getz(g, 0)]
-            else
-                error("getcoorddim($g) returned $ndim: expected 2 or 3")
-            end
-        elseif gtype in multipointtypes
-            Vector{Float64}[
-                GeoInterface.coordinates(getgeom(g, i - 1)) for i in 1:ngeom(g)
-            ]
-        elseif gtype in linetypes || gtype == wkbLinearRing
-            Vector{Float64}[
-                collect(getpoint(g, i - 1)[1:ndim]) for i in 1:ngeom(g)
-            ]
-        elseif gtype in multilinetypes || gtype in polygontypes
-            Vector{Vector{Float64}}[
-                GeoInterface.coordinates(getgeom(g, i - 1)) for i in 1:ngeom(g)
-            ]
-        elseif gtype in multipolygontypes
-            Vector{Vector{Vector{Float64}}}[
-                GeoInterface.coordinates(getgeom(g, i - 1)) for i in 1:ngeom(g)
-            ]
+    GeoInterface.is3d(::GeometryTraits, geom::AbstractGeometry) = is3d(geom)
+    function GeoInterface.ismeasured(::GeometryTraits, geom::AbstractGeometry)
+        return ismeasured(geom)
+    end
+
+    function GeoInterface.ncoord(::GeometryTraits, geom::AbstractGeometry)
+        return getcoorddim(geom)
+    end
+
+    function GeoInterface.getcoord(::GeometryTraits, geom::AbstractGeometry, i)
+        if i == 1
+            getx(geom, 0)
+        elseif i == 2
+            gety(geom, 0)
+        elseif i == 3 && is3d(geom)
+            getz(geom, 0)
+        elseif i == 3 && ismeasured(geom)
+            getm(geom, 0)
+        elseif i == 4 && ismeasured(geom) && is3d(geom)
+            getm(geom, 0)
+        else
+            return nothing
         end
     end
+
+    function GeoInterface.isempty(::GeometryTraits, geom::AbstractGeometry)
+        return isempty(geom)
+    end
+
+    function GeoInterface.ngeom(::GeometryTraits, geom::AbstractGeometry)
+        return ngeom(geom)
+    end
+
+    function GeoInterface.getgeom(
+        ::GeometryTraits,
+        geom::AbstractGeometry,
+        i::Integer,
+    )
+        return getgeom(geom, i - 1)
+    end
+
+    # Operations
```

Comment:
Can we add unit tests for the operations introduced from this line onwards? Other than that, I think it's fine that some of the new traits in `GeoInterface.geomtrait(...)` aren't being covered in this PR.

Reply:
Resolved by https://github.com/yeesian/ArchGDAL.jl/pull/290/commits/e2892d0d5a8d9a725c0a051ec173237f49e1638e